Add blockchain update mechanism, add request timeout to Peer

bfbachmann opened the pull request:
## Status
Ready for review

## Related Issue
#111 

## Description
* This adds a built-in timeout mechanism for peer requests. Now, when we call peer.Request and no response comes back after the default timeout period, we call the callback passed to peer.Request with a ProtocolError containing the RequestTimeout error code and message. This was users of that function don't have to implement their own timeout mechanisms as it is built into the protocol.
* Added `SyncBlockchain` function to iteratively request the next block from random peers. It is called when we boot with a target peer to connect to (download blockchain), and when we received a new block via a push message that is not valid wrt our blockchain. This way, if we are on the wrong fork of the blockchain, we miss a block, or we go offline for a while an then come back we will recognize that our blockchain is out of date and update it.
* A subtle but notable change: we now disconnect from a peer if we get 3 errors in reading it's messages over the entire course of our communication (they no longer need to be consecutive). 
* Pulled globals in App into the App struct
* Updated getLocalChain to just give us a new blockchain with a genisis block. Note that we always start with a genisis block, and if a target peer is provided we roll back our blockchain to an empty state and add the genisis block given to us by the peer.
* Added mining integration to App, as well as a `mine` flag. Mining is turned off my default. The miner restarts every time the blockchain is updated.
* Fixed a few bugs to do with decoding json messages
* Added miner command to the console for starting, stopping, and checking the status of the miner at will.
* Updated conn to use a better mock of net.Conn called BufConn 
* Fixed some bugs in Cloudbase transaction validation
* Added `BadRequest` (400) error code to message
* Broke peer testing into peer_test.go and peerstore_test.go

## WIKI Updates
* TODO

## Todos

General:
- [ ] Tests
- [ ] Wiki

bfbachmann committed:
Add blockchain update mechanism, add request timeout to Peer

bfbachmann committed:
Add Peer tests, remove Glide grabage

bfbachmann committed:
Merge branch 'dev' into 111-missed-blocks

bfbachmann committed:
Fix merge issue in dev

bfbachmann committed:
Update getLocalChain, update SyncBlockchain, refactor peer tests

chadlagore commented on app/app.go:
```diff
@@ -259,3 +271,92 @@ func (a *App) HandleBlock(blk *blockchain.Block) {
+
+// SyncBlockchain updates the local copy of the blockchain by requesting missing
+// blocks from peers. Returns error if we are not connected to any peers.
+func (a *App) SyncBlockchain() error {
```
Nitpick, lets use `SyncBlockChain`? (since the blockchain itself is pascal case)

chadlagore replied:
This is a huge function and very difficult to test. Is it possible to break it up a bit?

chadlagore commented on app/app.go:
```diff
@@ -204,11 +207,21 @@ func (a *App) PushHandler(push *msg.Push) {
+		Head:   blockchain.NilHash,
+	}
+
+	genisisBlock := blockchain.Genesis(user.HotWallet.Wallet.Public(),
```
Im not in love with the `User` API at the moment. I have some work to do there, this works for now.

chadlagore commented on app/app.go:
```diff
@@ -259,3 +271,92 @@ func (a *App) HandleBlock(blk *blockchain.Block) {
+		}
+
+		// Pick a peer to send the request to (this won't always be the same one)
+		p := a.PeerStore.Get(peerAddrs[0])
```
Without diving deep into the peer package, its very difficult to appreciate what this line is supposed to be doing. For example, how is it that this generates a different peer each loop? Shouldn't the addresses be abstracted away at this level? Why do I care which peer I'm connecting to (or what their address is)? I just want to send block requests out to the network.

bfbachmann replied:
Yea I agree, I think there should just be a `GetRandom` function for this. But we do need to select a peer to send the request to, because broadcasting a request doesn't really make much sense.

chadlagore commented on app/app.go:
```diff
@@ -259,3 +271,92 @@ func (a *App) HandleBlock(blk *blockchain.Block) {
+				continue
+			}
+
+			valid, validationCode := consensus.VerifyBlock(a.Chain, newBlock)
```
👏 thorough!

chadlagore commented on blockchain/block.go:
```diff
@@ -43,6 +43,16 @@ func (bh *BlockHeader) Marshal() []byte {
 
+// Equal returns true if all the fields (other than ExtraData) in each of
+// the BlockHeaders match, and false otherwise.
+func (bh *BlockHeader) Equal(otherHeader *BlockHeader) bool {
```
Reads nicer as `Equals`

bfbachmann replied:
I was going to make it `Equals` but that's not the Go way 🤷‍♂️.

chadlagore commented on blockchain/blockchain.go:
```diff
@@ -69,7 +69,7 @@ func (bc *BlockChain) ContainsTransaction(t *Transaction, start, stop uint32) (b
 }
 
-// CopyLocalBlockByIndex returns a copy of a block in the local chain by index.
+// CopyBlockByIndex returns a copy of a block in the local chain by index.
```
🤦‍♂️ 🙏

chadlagore commented on app/app.go:
```diff
@@ -259,3 +271,92 @@ func (a *App) HandleBlock(blk *blockchain.Block) {
+	errChan := make(chan *msg.ProtocolError)
+
+	// Define a handler for responses to our block requests
+	blockResponseHandler := func(blockResponse *msg.Response) {
```
This for example can be stubbed out for easy testing.

chadlagore commented on peer/peer.go:
```diff
@@ -159,11 +161,50 @@ func (p *Peer) Dispatch() {
 
 // Request sends the given request over this peer's Connection and registers the
 // given response hadnler to be called when the response arrives at the dispatcher.
```
handler* - realize its old stuff

chadlagore commented on peer/peerstore_test.go:
```diff
@@ -0,0 +1,253 @@
+package peer
```
In general, I think its a good idea to move towards the `testify` package and use assert statements. They really help with debugging, giving useful explanations of the failure (especially helpful when other people try to make changes in your code!)

chadlagore reviewed (commented):
Drive by comments. Looking good so far 🎉 

One general note about the approach we're using here to download the block chain, it seems like an unfortunately serial way to do it. I totally buy the idea of looking for the next block based on hash, but we basically do zero asynchronous torrenting from multiple peers like this (which was the goal I believe?)

bfbachmann committed:
Update blockchain,  App, console, tests, User, pool

Moved globals in the App package into the App struct.
Changed CopyBlockByIndex() to CopyBlockByLastBlockHash()
Change User to have Wallet instread of HotWallet
Added mining integration to App
Updated SyncBlockChain()
Added App tests
Updated tests and test_utils
Added miner command to console for checking and toggling miner status
Added mine flag to Cobra config

bfbachmann committed:
Add missed block detection to HandleBlock()

bfbachmann committed:
Update blockchain, tests, hash, cmd, app, consensus, message

App:
- Add blockchain download then target is specified
- Fix bugs in RequestHandler 
- Fix block decoding bugs in SyncBlockChain 

Blockchain:
- Add DecodeBlockJSON
- Change CopyBlockByBlockHash to GetBlockByBlockHash

Cmd/Conf:
- Change no-miner to miner, so we don't mine by default

Conn:
- Replace TestConn with BufConn

Consensus:
- Fix Cloudbase transaction validation bugs

Message:
- Add BadRequest protocol error
- Update decoding so it handles big numbers

bfbachmann committed:
Merge branch 'dev' into 111-missed-blocks

jordanschalm commented on app/app.go:
```diff
@@ -204,11 +231,19 @@ func (a *App) PushHandler(push *msg.Push) {
 
 // getLocalChain returns an instance of the blockchain.
-func getLocalChain() *blockchain.BlockChain {
+func getLocalChain(user *User) *blockchain.BlockChain {
```
Why is user a param?

jordanschalm replied:
It seems like the function is overloaded with two separate situations:
- we want to start a new chain with a genesis block
- we want to continue an old chain by restoring from storage

We should have two separate functions for these two situations.

jordanschalm commented on app/app.go:
```diff
@@ -204,11 +231,19 @@ func (a *App) PushHandler(push *msg.Push) {
+	genisisBlock := blockchain.Genesis(user.Wallet.Public(),
+		consensus.CurrentTarget(), consensus.StartingBlockReward, []byte{})
+
+	bc.AppendBlock(genisisBlock)
```
typo genisis -> genesis

bfbachmann replied:
Sad thing is I meant to write it like that... I thought that was how it was spelled :(.

jordanschalm commented on app/app.go:
```diff
@@ -236,26 +271,174 @@ func (a *App) HandleWork() {
+		chainChanged, err := a.SyncBlockChain()
+		if chainChanged && miner.IsMining() {
+			miner.StopMining()
+			go a.Mine()
```
worthwhile to have a `RestartMining` method? out of scope for this PR but make an issue for it if it's worth the time

jordanschalm commented on app/app.go:
```diff
@@ -236,26 +271,174 @@ func (a *App) HandleWork() {
+// blocks from peers. Returns true if the blockchain changed as a result of
+// calling this function, false if it didn't and an error if we are not connected
+// to any peers.
+func (a *App) SyncBlockChain() (bool, error) {
```
Can probably just have this return false if no peers and omit the error return
I'd also recommend checking for no peers before doing anything else and exiting early if we aren't connected to anything

bfbachmann replied:
It's possible, whoever unlikely, that all our peers disconnect from us while this function is running, in which case the blockchain might have been partially updated, but the update is incomplete. Checking if we are connected to peers before calling this would not guarantee that we have peers to make requests to throughout the duration of the function. So I think we still need both return values here.

jordanschalm commented on peer/peerstore.go:
```diff
@@ -156,12 +161,24 @@ func (ps *PeerStore) RemoveRandom() {
+func (ps *PeerStore) GetRandom() *Peer {
+	ps.lock.RLock()
+	defer ps.lock.RUnlock()
+	for _, p := range ps.peers {
```
lol

jordanschalm reviewed (changes requested):
Very nice work.

bfbachmann committed:
Update getLocalChain, update tests, add UpToDate for blockchain sync

chadlagore commented on app/app.go:
```diff
@@ -33,10 +33,13 @@ const (
+	Pool             *pool.Pool
+	blockQueue       chan *blockchain.Block
+	transactionQueue chan *blockchain.Transaction
+	quitChan         chan bool
```
:tada:

chadlagore commented on app/app.go:
```diff
@@ -80,6 +87,11 @@ func Run(cfg conf.Config) {
 	// stream in. Kick off a worker to handle requests and pushes.
 	go a.HandleWork()
 
+	if config.Mine {
```
This strikes me as a user-level configuration, but now I'm actually leaning towards one config file to rule them all. Perhaps we can move the `BlockSize` variable into `config` as well, then we can maybe do away with the user abstraction altogether 🤔 (Another PR)

bfbachmann replied:
The Cobra library we're using already supports config files by default.

chadlagore commented on app/app.go:
```diff
@@ -149,29 +169,41 @@ func (a *App) RequestHandler(req *msg.Request) msg.Response {
+		log.Debug("Received block request")
+
+		// Block is requested by block hash.
+		hashBytes, err := json.Marshal(req.Params["lastBlockHash"])
```
:tada:

chadlagore commented on app/app.go:
```diff
@@ -203,12 +235,20 @@ func (a *App) PushHandler(push *msg.Push) {
+// createBlockchain returns a new instance of a blockchain with only a genesis
+// block.
+func createBlockchain(user *User) *blockchain.BlockChain {
 	// around on the internets for one.
```
Might just blow this line away? :gun:

chadlagore commented on app/app.go:
```diff
@@ -236,26 +276,169 @@ func (a *App) HandleWork() {
+func (a *App) Mine() {
+	log.Info("Starting miner")
+	for {
+		// Make a new block form the transactions in the transaction pool
```
typo

chadlagore commented on app/app.go:
```diff
@@ -236,26 +276,169 @@ func (a *App) HandleWork() {
+			currentHeadHash = blockchain.HashSum(currentHead)
+		}
+
+		reqParams := map[string]interface{}{
```
A lot of the below seems like it should be its own function.
```go
requestNextBlock(currentHeadHash)
waitForBlock()
```

chadlagore commented on app/app.go:
```diff
@@ -236,26 +276,169 @@ func (a *App) HandleWork() {
+	changed := false
+
+	// Define a handler for responses to our block requests
+	blockResponseHandler := func(blockResponse *msg.Response) {
```
Nitpick - I'm sure its more performant to actually define this function. Then the go complier doesn't have to re-define and allocate space for this function every time we sync the chain (which seems like it will happen a lot) - it would also shrink this function body.

chadlagore commented on app/app.go:
```diff
@@ -236,26 +276,169 @@ func (a *App) HandleWork() {
+// blocks from peers. Returns true if the blockchain changed as a result of
+// calling this function, false if it didn't and an error if we are not connected
+// to any peers.
+func (a *App) SyncBlockChain() (bool, error) {
```
How would you test a function like this?

bfbachmann replied:
Hard to do, but it would have to be an integration tests where the App instance has a peerstore with connected peers who have copies of the blockchain. I don't see a way to make this function less complex without making it useless.

chadlagore replied:
Break it into 5 functions?

chadlagore commented on app/app.go:
```diff
@@ -236,26 +276,169 @@ func (a *App) HandleWork() {
+// create blocks, and mines those blocks. When a block is mined it is added
+// to the blockchain and broadcasted into the network. Mine() returns when
+// miner.StopMining() is called.
+func (a *App) Mine() {
```
Perhaps we should name this something else? Like `RunMiner`? Since we already have a fairly well defined `Mine` function.

chadlagore commented on app/app.go:
```diff
@@ -203,12 +235,20 @@ func (a *App) PushHandler(push *msg.Push) {
-	// TODO: Look for local chain on disk. If doesn't exist, go rummaging
+// createBlockchain returns a new instance of a blockchain with only a genesis
+// block.
+func createBlockchain(user *User) *blockchain.BlockChain {
```
Is this function just for testing? Why is the current user creating the genesis block in this block chain? `getLocalChain` actually performed a function that we can improve upon, `createBlockchain` seems to deliberately create an invalid blockchain.

bfbachmann replied:
Not invalid, just a genesis block, so it's valid. I got rid of `getLocalChain` because it was trying to do 2 things (create new blockchain or load one from disk). I'm breaking that into two functions, one is `createBlockchain` and the other will come when we support loading from disk.

chadlagore commented on app/app_test.go:
```diff
@@ -40,24 +42,23 @@ func TestPushHandlerNewTestTransaction(t *testing.T) {
-// func TestRequestHandlerNewBlockOK(t *testing.T) {
-// 	// Request a new block by hash and verify we get the right one.
-// 	a := createNewTestApp()
+func TestRequestHandlerNewBlockOK(t *testing.T) {
```
:tada:

chadlagore commented on app/app_test.go:
```diff
@@ -71,7 +72,7 @@ func TestRequestHandlerNewBlockBadParams(t *testing.T) {
 	// Make sure request failed.
 	assert.False(t, ok, "resource should not contain block")
-	assert.Equal(t, msg.ResourceNotFound, int(resp.Error.Code), resp.Error.Message)
+	assert.Equal(t, msg.BadRequest, int(resp.Error.Code), resp.Error.Message)
```
Thanks for making these legit error messages :+1:

chadlagore commented on app/console.go:
```diff
@@ -121,11 +129,41 @@ func connect(ctx *ishell.Context, a *App) {
+	emoji.Println(":mailbox: Address: " + wallet.Public().Repr())
+}
+
+func toggleMiner(ctx *ishell.Context, app *App) {
```
Cool!

chadlagore commented on app/test_utils.go:
```diff
@@ -7,9 +7,9 @@ import (
 )
 
-func createNewTestBlockRequest(lastBlock interface{}) *msg.Request {
+func createNewTestBlockRequest(lastBlockHash interface{}) *msg.Request {
```
:tada:

chadlagore commented on blockchain/block_test.go:
```diff
@@ -59,3 +58,28 @@ func TestBlockHeaderLen(t *testing.T) {
+	}
+
+	block2 := NewTestBlock()
+	if (&block1.BlockHeader).Equal(&block2.BlockHeader) {
```
We should use `assert`. Its less lines, easier to read, easier to get other people to work with your unit tests. Just far superior!

chadlagore commented on blockchain/blockchain.go:
```diff
@@ -69,14 +77,31 @@ func (bc *BlockChain) ContainsTransaction(t *Transaction, start, stop uint32) (b
-		b.Transactions = make([]*Transaction, len(blk.Transactions))
-		copy(b.Transactions, blk.Transactions)
-		return &b, nil
+// GetBlockByLastBlockHash returns a copy of the block in the local chain that
```
not a copy

chadlagore commented on blockchain/blockchain.go:
```diff
@@ -69,14 +77,31 @@ func (bc *BlockChain) ContainsTransaction(t *Transaction, start, stop uint32) (b
+
+// RollBack removes the last block from the blockchain. Returns the block that
+// was removed from the end of the chain, or nil if the blockchain is empty.
+func (bc *BlockChain) RollBack() *Block {
```
Interesting

chadlagore commented on blockchain/blockchain.go:
```diff
@@ -69,14 +77,31 @@ func (bc *BlockChain) ContainsTransaction(t *Transaction, start, stop uint32) (b
+		return nil
+	}
+	prevHead := bc.LastBlock()
+	bc.Blocks = bc.Blocks[:len(bc.Blocks)-1]
```
Fancy golang pop:

```
prevHead, bc.Blocks = bc.Blocks[len(bc.Blocks)-1], bc.Blocks[:len(bc.Blocks)-1]
```

https://github.com/golang/go/wiki/SliceTricks

chadlagore replied:
Yours is probably more readable 😆

chadlagore reviewed (commented):
Wow, lots going on here! Couple notes:

1. `assert` everywhere, literally saves you 50 lines in this pr and saves everyone else some serious head scratching. There is no reason not to use it at this point.

2. Long functions are hard to test, especially in the app package which already suffers from low coverage (it makes it really hard for others to make changes to your code in the future, the `Run` function is totally untested, and now the `SyncBlockChain` logic appears to be going the same way).

3. The scope is so far blown out, I don’t even see where this PR addresses ticket 111? Three separate PRs here would be better than one if you want legit feedback - its actually difficult to sign off on all of this because it takes so long to grok it all! Plus you expose yourself to too many TODO’s from onlookers. Takes forever to merge.

Theres lots about this PR that is sweet, but I :pray: to see two things before we merge: 

- [ ] Reduce the complexity of `SyncBlockChain` so that it doesn't go forever untested. Test it piece by piece.
- [ ] Use `assert`. It would take 10mins to swap these `t.FailNow`'s out.

**EDIT**: Im noticing that this is probably more geared towards issue 118? In that case the scope makes more sense! I thought this was literally random coding lol.

bfbachmann committed:
Refactor SyncBlockChain, update/add tests, change Mine() -> RunMiner()

chadlagore commented on app/app.go:
```diff
@@ -343,15 +338,22 @@ func (a *App) Mine() {
 }
 
+// RestartMiner restarts the miner only if it is running when this is called.
+func (a *App) RestartMiner() {
```
🎉

chadlagore reviewed (approved):
Looks great!

bfbachmann committed:
Merge branch 'dev' into 111-missed-blocks

bfbachmann committed:
Add target to new blocks so we can mine

bfbachmann committed:
Clean up app, fix bug in peer handshake
